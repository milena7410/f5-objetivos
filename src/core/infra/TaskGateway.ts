import type { Task, TaskDTO } from "../domain/Task";

export interface TaskGateway {
  getTasks(): Promise<Task[]>;
  getTask(id: number): Promise<Task>;
  createTask(todo: TaskDTO): Promise<Task>;
  completeTask(task: Task): Promise<Task>;
  undoCompletedTask(task: Task): Promise<Task>;
  deleteTask(id: number): Promise<void>;
  deleteAllTasks(): Promise<any[]>;
}
