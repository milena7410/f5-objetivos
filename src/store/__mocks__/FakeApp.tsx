import * as React from "react";
import {
  DarkTheme,
  DefaultTheme,
  ThemeProvider,
} from "@react-navigation/native";
import { Text } from "react-native";

import { ThemedView } from "~/components/ThemedView";
import { useTodos } from "~/store/reducers/todos/actions";
import { useThemeContext } from "~/contexts/ThemeContext";

export const GetTodoListEmpty = () => {
  const { todos } = useTodos();
  const { colorScheme } = useThemeContext();
  return (
    <ThemeProvider value={colorScheme === "dark" ? DarkTheme : DefaultTheme}>
      <ThemedView lightColor="green" darkColor="gray">
        {todos.list.map((todo) => (
          <Text key={todo.id}>{todo.title}</Text>
        ))}
      </ThemedView>
    </ThemeProvider>
  );
};

export const GetTodoList = () => {
  const { getTodoList, todos } = useTodos();
  React.useEffect(() => {
    getTodoList();
  }, []);
  return (
    <ThemedView>
      {todos.list.map((todo) => (
        <Text key={todo.id}>{todo.title}</Text>
      ))}
    </ThemedView>
  );
};

export const CreateTodo = () => {
  const { addTodo, todos } = useTodos();
  React.useEffect(() => {
    addTodo({ title: "New TODO", userId: 211 });
  }, []);
  return (
    <ThemedView>
      {todos.list.map((todo) => (
        <Text key={todo.id}>{todo.title}</Text>
      ))}
    </ThemedView>
  );
};

export const NotFound = ({ id }: { id: number }) => {
  const { todos, getTodo } = useTodos();
  React.useEffect(() => {
    getTodo(id);
  }, [id]);
  return <ThemedView>{!!todos.error && <Text>{todos.error}</Text>}</ThemedView>;
};

export const DeleteTask = ({ id }: { id: number }) => {
  const { todos, deleteTask, getTodoList } = useTodos();
  const first = React.useRef(true);
  React.useEffect(() => {
    getTodoList().finally(() =>
      deleteTask(id)
        .unwrap()
        .catch(() => {})
    );
  }, [id]);

  return (
    <ThemedView>
      {todos.list.map((todo) => (
        <Text key={todo.id}>{todo.title}</Text>
      ))}
      {!!todos.error && <Text>{todos.error}</Text>}
    </ThemedView>
  );
};

export const CompleteTask = ({
  id,
  method,
}: {
  id: number;
  method: "completeTask" | "undoCompletedTask";
}) => {
  const { todos, getTodoList, ...todoHook } = useTodos();
  React.useEffect(() => {
    getTodoList().finally(() =>
      todoHook[method](id)
        .unwrap()
        .catch(() => {})
    );
  }, [id]);
  const { colorScheme } = useThemeContext();
  const task = todos.list.find((task) => task.id === id);

  return (
    <ThemeProvider value={colorScheme === "dark" ? DarkTheme : DefaultTheme}>
      <ThemedView lightColor="green" darkColor="gray">
        <Text>{task?.completed ? "COMPLETED" : "NOT_COMPLETED"}</Text>
        {!!todos.error && <Text>{todos.error}</Text>}
      </ThemedView>
    </ThemeProvider>
  );
};
