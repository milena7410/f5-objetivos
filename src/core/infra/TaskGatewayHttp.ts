import { HttpClient } from "../domain/HttpClient";
import { taskBuilder, TaskDTO, Task } from "../domain/Task";
import { TaskGateway } from "./TaskGateway";

export function tasksGatewayHttp(httpClient: HttpClient): TaskGateway {
  const getTasks = async () => {
    const todoList = await httpClient.get<TaskDTO[]>("/todos");
    const tasks = todoList.map((task) => taskBuilder(task));
    return tasks;
  };

  const getTask = async (id: number) => {
    const newTask = await httpClient.get<TaskDTO>(`/todos/${id}`);
    return taskBuilder(newTask);
  };

  const createTask = async (todo: TaskDTO) => {
    const newTask = await httpClient.post<TaskDTO>("/todos", todo);
    return taskBuilder(newTask);
  };

  const deleteTask = async (id: number) => {
    await httpClient.delete<TaskDTO>(`/todos/${id}`);
  };

  const editTask = async (task: Task) => {
    const updatedTask = await httpClient.put<Task>(`/todos/${task.id}`, task);
    return taskBuilder(updatedTask);
  };

  const completeTask = async (task: Task) => {
    const updatedTask = await httpClient.put<Task>(
      `/todos/${task.id}/complete`,
      task
    );
    return taskBuilder(updatedTask);
  };

  const undoCompletedTask = async (task: Task) => {
    const updatedTask = await httpClient.put<Task>(
      `/todos/${task.id}/complete`,
      task
    );
    return taskBuilder(updatedTask);
  };

  const deleteAllTasks = async () => {
    await httpClient.delete(`/todos/all`);
    return [];
  };

  return {
    getTasks,
    createTask,
    getTask,
    editTask,
    deleteTask,
    completeTask,
    undoCompletedTask,
    deleteAllTasks,
  };
}
