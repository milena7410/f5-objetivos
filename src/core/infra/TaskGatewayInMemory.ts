import { Task, taskBuilder, TaskDTO } from "../domain/Task";
import { TaskGateway } from "./TaskGateway";

export const TODO_LIST_MOCK = [
  {
    userId: 10,
    id: 191,
    title:
      "temporibus atque distinctio omnis eius impedit tempore molestias pariatur",
    completed: false,
  },
  {
    userId: 10,
    id: 192,
    title: "ut quas possimus exercitationem sint voluptates",
    completed: true,
  },
  {
    userId: 10,
    id: 193,
    title: "rerum debitis voluptatem qui eveniet tempora distinctio a",
    completed: true,
  },
  {
    userId: 10,
    id: 194,
    title: "sed ut vero sit molestiae",
    completed: false,
  },
  {
    userId: 10,
    id: 195,
    title: "rerum ex veniam mollitia voluptatibus pariatur",
    completed: true,
  },
  {
    userId: 10,
    id: 196,
    title:
      "consequuntur aut ut fugit similique dignissimos quo nobis earum saepe quis eius est sint explicabo numquam repellendus a magnam ",
    completed: true,
  },
  {
    userId: 10,
    id: 197,
    title: "dignissimos quo nobis earum saepe",
    completed: true,
  },
  {
    userId: 10,
    id: 198,
    title: "quis eius est sint explicabo",
    completed: true,
  },
  {
    userId: 10,
    id: 199,
    title: "numquam repellendus a magnam",
    completed: true,
  },
  {
    userId: 10,
    id: 200,
    title: "ipsam aperiam voluptates qui",
    completed: false,
  },
];
export const TaskGatewayInMemory = (): TaskGateway => {
  const TODO_LIST: Task[] = JSON.parse(JSON.stringify(TODO_LIST_MOCK));
  const getTasks = async (): Promise<Task[]> => TODO_LIST.map(taskBuilder);

  const createTask = async (task: TaskDTO): Promise<Task> => {
    const newTask = taskBuilder(task);
    TODO_LIST.push(newTask);
    return newTask;
  };

  const getTask = async (id: number): Promise<Task> => {
    const task = TODO_LIST.find((t) => t.id === id);
    if (!task) {
      throw new Error("not found");
    }
    return taskBuilder(task);
  };

  const deleteTask = async (id: number) => {
    const taskIndex = TODO_LIST.findIndex((t) => t.id === id);
    if (taskIndex < 0) {
      throw new Error("not found");
    }
    TODO_LIST.splice(taskIndex, 1);
  };

  const completeTask = async (task: Task) => {
    const currentTask = TODO_LIST.find((t) => t.id === task.id);
    if (!currentTask) {
      throw new Error("not found");
    }
    return taskBuilder({ ...currentTask, completed: true });
  };

  const undoCompletedTask = async (task: Task) => {
    const currentTask = TODO_LIST.find((t) => t.id === task.id);
    if (!currentTask) {
      throw new Error("not found");
    }
    return taskBuilder({ ...currentTask, completed: false });
  };

  return {
    getTasks,
    createTask,
    getTask,
    deleteTask,
    completeTask,
    undoCompletedTask,
  };
};
