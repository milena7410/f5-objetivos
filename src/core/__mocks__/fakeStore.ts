import { configureStore } from "@reduxjs/toolkit";

import { rootReducer } from "../../store/reducers";
import { TaskGatewayInMemory } from "../infra/TaskGatewayInMemory";

export const store = () => {
  return configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        thunk: { extraArgument: { taskGateway: TaskGatewayInMemory() } },
      }),
  });
};
