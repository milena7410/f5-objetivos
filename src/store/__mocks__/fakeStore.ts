import { configureStore } from "@reduxjs/toolkit";

import { rootReducer } from "../reducers";
import { TaskGatewayInMemory } from "../../core/infra/TaskGatewayInMemory";

export const store = () => {
  return configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        thunk: { extraArgument: { taskGateway: TaskGatewayInMemory() } },
      }),
  });
};
