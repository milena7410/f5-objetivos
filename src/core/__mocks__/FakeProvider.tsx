import * as React from "react";
import {
  Provider,
  TypedUseSelectorHook,
  useDispatch,
  useSelector,
} from "react-redux";
import { store } from "./fakeStore";
import { AppDispatch, RootState } from "../../store";

// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

export const FakeReduxProvider = ({ children }: React.PropsWithChildren) => {
  return <Provider store={store()}>{children}</Provider>;
};
