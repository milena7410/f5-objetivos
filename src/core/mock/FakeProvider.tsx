import {
  Provider,
  TypedUseSelectorHook,
  useDispatch,
  useSelector,
} from "react-redux";
import { store } from "./fakeStore";
import { AppDispatch, RootState } from "../store";
import { PropsWithChildren } from "react";

// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

export const FakeReduxProvider = ({ children }: PropsWithChildren) => {
  return <Provider store={store()}>{children}</Provider>;
};
