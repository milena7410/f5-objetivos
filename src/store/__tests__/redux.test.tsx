import * as React from "react";
import {
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react-native";

import { FakeReduxProvider } from "../__mocks__/FakeProvider";
import { TODO_LIST_MOCK } from "~/core/infra/TaskGatewayInMemory";
import {
  CompleteTask,
  CreateTodo,
  DeleteTask,
  EditTask,
  GetTodoList,
  GetTodoListEmpty,
  NotFound,
} from "../__mocks__/FakeApp";

describe("Redux TodoList", () => {
  const [FIRST_TASK, SECOND_TASK_COMPLETED] = TODO_LIST_MOCK;
  it("should render a list of todos correctly", async () => {
    render(
      <FakeReduxProvider>
        <GetTodoList />
      </FakeReduxProvider>
    );
    await waitFor(() => {
      expect(screen.getByText(FIRST_TASK.title)).toBeTruthy();
      expect(screen.toJSON()).toMatchSnapshot();
    });
  });

  it("should not render anything if the todo list is empty", () => {
    const { queryByRole, toJSON } = render(
      <FakeReduxProvider>
        <GetTodoListEmpty />
      </FakeReduxProvider>
    );
    expect(toJSON()).toMatchSnapshot();
    expect(queryByRole("text")).toBeFalsy();
  });

  it("should create a new Task and render", async () => {
    render(
      <FakeReduxProvider>
        <CreateTodo />
      </FakeReduxProvider>
    );
    await waitFor(() => {
      expect(screen.getByText("New TODO")).toBeTruthy();
    });
  });

  it("should throw error if not exists and render error label", async () => {
    render(
      <FakeReduxProvider>
        <NotFound id={0} />
      </FakeReduxProvider>
    );
    await waitFor(() => {
      expect(screen.getByText("not found")).toBeTruthy();
    });
  });

  it("should throw error if try delete task not exists", async () => {
    render(
      <FakeReduxProvider>
        <DeleteTask id={0} />
      </FakeReduxProvider>
    );
    await waitFor(() => {
      expect(screen.getByText(FIRST_TASK.title)).toBeTruthy();
      expect(screen.getByText("not found")).toBeTruthy();
    });
  });

  it("should delete task", async () => {
    render(
      <FakeReduxProvider>
        <DeleteTask id={FIRST_TASK.id} />
      </FakeReduxProvider>
    );

    await waitFor(() => {
      expect(screen.queryByText(FIRST_TASK.title)).toBeFalsy();
      expect(screen.queryByText("not found")).toBeFalsy();
    });
  });

  it("should complete task", async () => {
    render(
      <FakeReduxProvider>
        <CompleteTask method="completeTask" id={FIRST_TASK.id} />
      </FakeReduxProvider>
    );

    await waitFor(() => {
      expect(screen.getByText("COMPLETED")).toBeTruthy();
      expect(screen.queryByText("task already completed")).toBeFalsy();
    });
  });

  it("should throw error if try complete a completed task", async () => {
    render(
      <FakeReduxProvider>
        <CompleteTask method="completeTask" id={SECOND_TASK_COMPLETED.id} />
      </FakeReduxProvider>
    );

    await waitFor(() => {
      expect(screen.getByText("COMPLETED")).toBeTruthy();
      expect(screen.getByText("task already completed")).toBeTruthy();
    });
  });

  it("should undo completed task", async () => {
    render(
      <FakeReduxProvider>
        <CompleteTask
          method="undoCompletedTask"
          id={SECOND_TASK_COMPLETED.id}
        />
      </FakeReduxProvider>
    );

    await waitFor(() => {
      expect(screen.getByText("NOT_COMPLETED")).toBeTruthy();
      expect(screen.queryByText("task not completed")).toBeFalsy();
    });
  });

  it("should throw error when it tries complete a completed task", async () => {
    render(
      <FakeReduxProvider>
        <CompleteTask method="undoCompletedTask" id={FIRST_TASK.id} />
      </FakeReduxProvider>
    );

    await waitFor(() => {
      expect(screen.getByText("NOT_COMPLETED")).toBeTruthy();
      expect(screen.getByText("task not completed")).toBeTruthy();
    });
  });
  it("should throw an error when it tries to manage the attribute 'completed' that is not found.", async () => {
    render(
      <FakeReduxProvider>
        <CompleteTask method="completeTask" id={1} />
        <CompleteTask method="undoCompletedTask" id={1} />
      </FakeReduxProvider>
    );

    await waitFor(() => {
      expect(screen.getAllByText("not found").length).toBe(2);
    });
  });

  it("should delete all tasks", async () => {
    render(
      <FakeReduxProvider>
        <GetTodoList />
      </FakeReduxProvider>
    );
    await waitFor(() => {
      expect(screen.getByText(FIRST_TASK.title)).toBeTruthy();
    });
    await waitFor(() => {
      const buttonDeleteAll = screen.getByText("DELETE ALL");
      fireEvent(buttonDeleteAll, "press");
      expect(screen.queryByText(FIRST_TASK.title)).toBeFalsy();
    });
  });

  it("should edit task", async () => {
    const newTitle = "TASK EDITED";
    render(
      <FakeReduxProvider>
        <EditTask title={newTitle} id={FIRST_TASK.id} />
      </FakeReduxProvider>
    );

    await waitFor(() => {
      expect(screen.getByText(FIRST_TASK.title)).toBeTruthy();
      expect(screen.queryByText(newTitle)).toBeFalsy();
      const buttonDeleteAll = screen.getByText("EDIT");
      fireEvent(buttonDeleteAll, "press");
    });
    await waitFor(() => {
      expect(screen.queryByText(FIRST_TASK.title)).toBeFalsy();
      expect(screen.getByText(newTitle)).toBeTruthy();
    });
  });
});
