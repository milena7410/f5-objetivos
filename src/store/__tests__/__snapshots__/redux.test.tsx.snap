// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux TodoList should not render anything if the todo list is empty 1`] = `
<View
  className=" bg-white dark:bg-black "
/>
`;

exports[`Redux TodoList should render a list of todos correctly 1`] = `
<View
  className=" bg-white dark:bg-black "
>
  <Text>
    temporibus atque distinctio omnis eius impedit tempore molestias pariatur
  </Text>
  <Text>
    ut quas possimus exercitationem sint voluptates
  </Text>
  <Text>
    rerum debitis voluptatem qui eveniet tempora distinctio a
  </Text>
  <Text>
    sed ut vero sit molestiae
  </Text>
  <Text>
    rerum ex veniam mollitia voluptatibus pariatur
  </Text>
  <Text>
    consequuntur aut ut fugit similique dignissimos quo nobis earum saepe quis eius est sint explicabo numquam repellendus a magnam 
  </Text>
  <Text>
    dignissimos quo nobis earum saepe
  </Text>
  <Text>
    quis eius est sint explicabo
  </Text>
  <Text>
    numquam repellendus a magnam
  </Text>
  <Text>
    ipsam aperiam voluptates qui
  </Text>
</View>
`;
